fix: setting 64bit fields from strings supported

software-dov opened the pull request:
Due to limitations in certain browser/javascript combinations,
fields that are 64 bit integer types are converted to strings when
encoding a message to JSON or a dict.
Decoding from JSON handles this explicitly, but it makes converting
back from a dict an error.

This fix adds support for setting these 64 bit fields explicitly from
strings.

E.g.
```py
class Squid(proto.Message):
      mass_kg = proto.Field(proto.INT64, number=1)

s = Squid(mass_kg=10)
s_dict = Squid.to_dict(s)

assert s == Squid(s_dict)
s.mass_kg = "20"            # Supported
```

software-dov committed:
fix: setting 64bit fields from strings supported

Due to limitations in certain browser/javascript combinations,
fields that are 64 bit integer types are converted to strings when
encoding a message to JSON or a dict.
Decoding from JSON handles this explicitly, but it makes converting
back from a dict an error.

This fix adds support for setting these 64 bit fields explicitly from
strings.

E.g.

class Squid(proto.Message):
      mass_kg = proto.Field(proto.INT64, number=1)

s = Squid(mass_kg=10)
s_dict = Squid.to_dict(s)

assert s == Squid(**s_dict) # Supported
assert s == Squid(s_dict)   # NOT supported for performance reasons
s.mass_kg = "20"            # Supported

software-dov commented on docs/marshal.rst:
```diff
@@ -72,7 +72,28 @@ Protocol buffer type                Python type             Nullable
+
+      msg_2 = MyMessage(msg_dict)  # Raises an exception
+
+      msg_3 = MyMessage(**msg_dict) # Works without exception
```
I'm deeply conflicted about that.

On the one hand, there is a definite performance advantage to be had in *not* unpacking the dict in the cases where there is no 64 bit field. Any conversions and marshaling is handled by the protobuf in c++. Messages with 64 bit fields seem to be fairly rare if this is the first we've encountered the issue, and especially for large messages the performance impact is likely to be non-trivial.

On the other hand, most people are not going to know or care about this particular bug until their code breaks, and then they'll be frustrated that there's this special and opaque seeming workaround/corner case. There is a definite argument to be made for consistency over raw performance.

On the third hand, this round-trip-dict shenanigans is a niche use case that we use for doing path parameter expansion with REST transports. Most of the time, just converting from dicts normally should be fine.

tseaver replied:
I'd honestly rather have a flag passed in than have the `**` unpacking be the suggested workaround -- that is a brain-bender for anybody not already deep into the weeds of our implementation.

We should be asking for the C++ guys to fix this upstream, as well:  the round-trip should Just Work(TM).

tseaver commented on docs/marshal.rst:
```diff
@@ -72,7 +72,28 @@ Protocol buffer type                Python type             Nullable
 
+   Due to certain browser/javascript limitations, 64 bit sized fields, e.g. INT64, UINT64,
+   are converted to strings when marshalling messages to dictionaries or JSON.
+   Decoding JSON handles this correctly, but dicts must be unpacked when reconstructing messages. This is necessary to trigger a special case workaround.
```
```suggestion
   Decoding JSON handles this correctly, but dicts must be unpacked when reconstructing messages.
   This is necessary to trigger a special case workaround.
```

software-dov committed:
Update docs/marshal.rst

Co-authored-by: Tres Seaver <tseaver@palladion.com>

software-dov committed:
Integrate reviews and handle deeply nested messages

Add test and impl for deeply nested messages with 64 bit int fields
Remove doc changes

software-dov committed:
Merge branch 'stringy-int64' of https://github.com/software-dov/proto-plus-python into stringy-int64

software-dov committed:
Satiate stylecheck
